File: C/count_min_sketch/count_min_sketch.c
#include "count_min_sketch.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdint.h>

#define FNV_PRIME 0x100000001b3ULL

static uint64_t fnv_hash(const char* data, size_t len, uint64_t seed) {
    uint64_t hash = 0xcbf29ce484222325ULL ^ seed;
    for (size_t i = 0; i < len; i++) {
        hash = hash * FNV_PRIME + (uint8_t)data[i];
    }
    return hash;
}

static void get_hashes(const cms_sketch_t* sketch, const char* item, size_t len, uint64_t* out) {
    uint64_t h1 = fnv_hash(item, len, 0);
    uint64_t h2 = fnv_hash(item, len, sketch->seed);
    
    for (size_t i = 0; i < sketch->depth; i++) {
        out[i] = h1 + i * h2;
    }
}

cms_sketch_t* cms_new(size_t depth, size_t width) {
    if (depth < 1) depth = 1;
    if (width < 2) width = 2;
    
    cms_sketch_t* sketch = (cms_sketch_t*)malloc(sizeof(cms_sketch_t));
    if (!sketch) return NULL;
    
    sketch->depth = depth;
    sketch->width = width;
    sketch->seed = 0xDEADBEEF;
    sketch->total_count = 0;
    
    sketch->table = (uint64_t**)malloc(depth * sizeof(uint64_t*));
    if (!sketch->table) {
        free(sketch);
        return NULL;
    }
    
    for (size_t i = 0; i < depth; i++) {
        sketch->table[i] = (uint64_t*)calloc(width, sizeof(uint64_t));
        if (!sketch->table[i]) {
            for (size_t j = 0; j < i; j++) free(sketch->table[j]);
            free(sketch->table);
            free(sketch);
            return NULL;
        }
    }
    
    return sketch;
}

cms_sketch_t* cms_new_default(void) {
    return cms_new(10, 1000);
}

void cms_free(cms_sketch_t* sketch) {
    if (!sketch) return;
    for (size_t i = 0; i < sketch->depth; i++) {
        free(sketch->table[i]);
    }
    free(sketch->table);
    free(sketch);
}

void cms_update(cms_sketch_t* sketch, const char* item, size_t item_len, uint64_t delta) {
    uint64_t hashes[64];  // max depth
    get_hashes(sketch, item, item_len, hashes);
    
    for (size_t i = 0; i < sketch->depth; i++) {
        size_t idx = hashes[i] % sketch->width;
        sketch->table[i][idx] += delta;
    }
    sketch->total_count += delta;
}

void cms_increment(cms_sketch_t* sketch, const char* item, size_t item_len) {
    cms_update(sketch, item, item_len, 1);
}

uint64_t cms_estimate(const cms_sketch_t* sketch, const char* item, size_t item_len) {
    uint64_t hashes[64];
    get_hashes(sketch, item, item_len, hashes);
    
    uint64_t min_val = sketch->table[0][hashes[0] % sketch->width];
    for (size_t i = 1; i < sketch->depth; i++) {
        size_t idx = hashes[i] % sketch->width;
        uint64_t val = sketch->table[i][idx];
        if (val < min_val) min_val = val;
    }
    return min_val;
}

uint64_t cms_total_count(const cms_sketch_t* sketch) {
    return sketch->total_count;
}

int cms_merge(cms_sketch_t* dest, const cms_sketch_t* src) {
    if (dest->depth != src->depth || dest->width != src->width) {
        return CMS_ERR_DIM_MISMATCH;
    }
    
    for (size_t i = 0; i < dest->depth; i++) {
        for (size_t j = 0; j < dest->width; j++) {
            dest->table[i][j] += src->table[i][j];
        }
    }
    dest->total_count += src->total_count;
    return CMS_OK;
}

uint8_t* cms_to_bytes(const cms_sketch_t* sketch, size_t* out_len) {
    size_t size = 32 + sketch->depth * sketch->width * 8;
    uint8_t* bytes = (uint8_t*)malloc(size);
    if (!bytes) return NULL;
    
    memset(bytes, 0, size);
    
    size_t offset = 0;
    
    // depth (8 bytes)
    for (int i = 0; i < 8; i++) bytes[offset++] = (sketch->depth >> (i * 8)) & 0xFF;
    // width (8 bytes)
    for (int i = 0; i < 8; i++) bytes[offset++] = (sketch->width >> (i * 8)) & 0xFF;
    // seed (8 bytes)
    for (int i = 0; i < 8; i++) bytes[offset++] = (sketch->seed >> (i * 8)) & 0xFF;
    // total_count (8 bytes)
    for (int i = 0; i < 8; i++) bytes[offset++] = (sketch->total_count >> (i * 8)) & 0xFF;
    
    for (size_t i = 0; i < sketch->depth; i++) {
        for (size_t j = 0; j < sketch->width; j++) {
            uint64_t val = sketch->table[i][j];
            for (int k = 0; k < 8; k++) bytes[offset++] = (val >> (k * 8)) & 0xFF;
        }
    }
    
    *out_len = size;
    return bytes;
}

cms_sketch_t* cms_from_bytes(const uint8_t* bytes, size_t len) {
    if (len < 32) return NULL;
    
    size_t offset = 0;
    size_t depth = 0, width = 0;
    uint64_t seed = 0, total_count = 0;
    
    for (int i = 0; i < 8; i++) depth |= ((size_t)bytes[offset++]) << (i * 8);
    for (int i = 0; i < 8; i++) width |= ((size_t)bytes[offset++]) << (i * 8);
    for (int i = 0; i < 8; i++) seed |= ((uint64_t)bytes[offset++]) << (i * 8);
    for (int i = 0; i < 8; i++) total_count |= ((uint64_t)bytes[offset++]) << (i * 8);
    
    size_t expected = 32 + depth * width * 8;
    if (len < expected) return NULL;
    
    cms_sketch_t* sketch = cms_new(depth, width);
    if (!sketch) return NULL;
    
    sketch->seed = seed;
    sketch->total_count = total_count;
    
    for (size_t i = 0; i < depth; i++) {
        for (size_t j = 0; j < width; j++) {
            uint64_t val = 0;
            for (int k = 0; k < 8; k++) val |= ((uint64_t)bytes[offset++]) << (k * 8);
            sketch->table[i][j] = val;
        }
    }
    
    return sketch;
}

void cms_clear(cms_sketch_t* sketch) {
    for (size_t i = 0; i < sketch->depth; i++) {
        memset(sketch->table[i], 0, sketch->width * sizeof(uint64_t));
    }
    sketch->total_count = 0;
}

cms_config_t cms_optimal(double epsilon, double delta) {
    cms_config_t config;
    config.width = (size_t)ceil(exp(1.0) / epsilon);
    config.depth = (size_t)ceil(-log(delta));
    config.seed = 0xDEADBEEF;
    if (config.depth < 1) config.depth = 1;
    if (config.width < 2) config.width = 2;
    return config;
}

cms_sketch_t* cms_with_rate(double epsilon, double delta) {
    cms_config_t config = cms_optimal(epsilon, delta);
    return cms_new(config.depth, config.width);
}

void cms_dimensions(const cms_sketch_t* sketch, size_t* depth, size_t* width) {
    if (depth) *depth = sketch->depth;
    if (width) *width = sketch->width;
}

File: Go/count_min_sketch/count_min_sketch.go
// Package count_min_sketch implements a probabilistic data structure for frequency estimation.
// Uses sublinear space for counting item frequencies with controlled error bounds.
package count_min_sketch

import (
	"encoding/binary"
	"hash/fnv"
	"math"
)

// Config holds configuration for Count-Min Sketch
type Config struct {
	Depth uint
	Width uint
	Seed  uint64
}

// CountMinSketch represents a Count-Min Sketch data structure
type CountMinSketch struct {
	table       [][]uint64
	depth       uint
	width       uint
	seed        uint64
	totalCount  uint64
}

// New creates a new Count-Min Sketch with given configuration
func New(depth, width uint) *CountMinSketch {
	if depth < 1 {
		depth = 1
	}
	if width < 2 {
		width = 2
	}
	
	table := make([][]uint64, depth)
	for i := range table {
		table[i] = make([]uint64, width)
	}
	
	return &CountMinSketch{
		table:      table,
		depth:      depth,
		width:      width,
		seed:       0xDEADBEEF,
		totalCount: 0,
	}
}

// NewWithConfig creates a Count-Min Sketch from configuration
func NewWithConfig(config Config) *CountMinSketch {
	return New(config.Depth, config.Width)
}

// Optimal creates configuration based on desired error rate and confidence
// epsilon: relative error bound (e.g., 0.01 = 1% error)
// delta: confidence probability (e.g., 0.01 = 99% confidence)
func Optimal(epsilon, delta float64) Config {
	width := math.Ceil(math.E / epsilon)
	depth := math.Ceil(-math.Log(delta))

	return Config{
		Depth: uint(math.Max(1, depth)),
		Width: uint(math.Max(2, width)),
		Seed:  0xDEADBEEF,
	}
}

// WithRate creates a new sketch with optimal config for given epsilon and delta
func WithRate(epsilon, delta float64) *CountMinSketch {
	config := Optimal(epsilon, delta)
	return New(config.Depth, config.Width)
}

// Update increases the count for an item by delta
func (c *CountMinSketch) Update(item string, delta uint64) {
	hashes := c.getHashes(item)
	for i, h := range hashes {
		idx := h % uint64(c.width)
		c.table[i][idx] += delta
	}
	c.totalCount += delta
}

// Increment increases count by 1
func (c *CountMinSketch) Increment(item string) {
	c.Update(item, 1)
}

// Estimate returns the estimated count for an item (upper bound)
func (c *CountMinSketch) Estimate(item string) uint64 {
	hashes := c.getHashes(item)
	min := uint64(0)
	for i, h := range hashes {
		idx := h % uint64(c.width)
		val := c.table[i][idx]
		if i == 0 || val < min {
			min = val
		}
	}
	return min
}

// TotalCount returns total number of items processed
func (c *CountMinSketch) TotalCount() uint64 {
	return c.totalCount
}

// Dimensions returns (depth, width)
func (c *CountMinSketch) Dimensions() (uint, uint) {
	return c.depth, c.width
}

// Merge combines another sketch into this one (must have same dimensions)
func (c *CountMinSketch) Merge(other *CountMinSketch) error {
	if c.depth != other.depth || c.width != other.width {
		return ErrDimensionMismatch
	}
	for i := uint(0); i < c.depth; i++ {
		for j := uint(0); j < c.width; j++ {
			c.table[i][j] += other.table[i][j]
		}
	}
	c.totalCount += other.totalCount
	return nil
}

// ToBytes serializes the sketch to bytes
func (c *CountMinSketch) ToBytes() []byte {
	totalSize := 32 + int(c.depth)*int(c.width)*8
	bytes := make([]byte, totalSize)
	
	binary.LittleEndian.PutUint64(bytes[0:8], uint64(c.depth))
	binary.LittleEndian.PutUint64(bytes[8:16], uint64(c.width))
	binary.LittleEndian.PutUint64(bytes[16:24], c.seed)
	binary.LittleEndian.PutUint64(bytes[24:32], c.totalCount)
	
	offset := 32
	for i := uint(0); i < c.depth; i++ {
		for j := uint(0); j < c.width; j++ {
			binary.LittleEndian.PutUint64(bytes[offset:offset+8], c.table[i][j])
			offset += 8
		}
	}
	
	return bytes
}

// FromBytes deserializes a sketch from bytes
func FromBytes(bytes []byte) (*CountMinSketch, error) {
	if len(bytes) < 32 {
		return nil, ErrTooShort
	}
	
	depth := binary.LittleEndian.Uint64(bytes[0:8])
	width := binary.LittleEndian.Uint64(bytes[8:16])
	seed := binary.LittleEndian.Uint64(bytes[16:24])
	totalCount := binary.LittleEndian.Uint64(bytes[24:32])
	
	expectedLen := 32 + int(depth)*int(width)*8
	if len(bytes) < expectedLen {
		return nil, ErrInvalidLength
	}
	
	table := make([][]uint64, depth)
	offset := 32
	for i := uint64(0); i < depth; i++ {
		table[i] = make([]uint64, width)
		for j := uint64(0); j < width; j++ {
			table[i][j] = binary.LittleEndian.Uint64(bytes[offset : offset+8])
			offset += 8
		}
	}
	
	return &CountMinSketch{
		table:      table,
		depth:      uint(depth),
		width:      uint(width),
		seed:       seed,
		totalCount: totalCount,
	}, nil
}

// Clear resets the sketch
func (c *CountMinSketch) Clear() {
	for i := range c.table {
		for j := range c.table[i] {
			c.table[i][j] = 0
		}
	}
	c.totalCount = 0
}

func (c *CountMinSketch) getHashes(item string) []uint64 {
	h1 := fnv.New64a()
	h1.Write([]byte(item))
	sum1 := h1.Sum64()
	
	h2 := fnv.New64a()
	seedBytes := []byte{
		byte(c.seed >> 56), byte(c.seed >> 48), byte(c.seed >> 40), byte(c.seed >> 32),
		byte(c.seed >> 24), byte(c.seed >> 16), byte(c.seed >> 8), byte(c.seed),
	}
	h2.Write(seedBytes)
	h2.Write([]byte(item))
	sum2 := h2.Sum64()
	
	hashes := make([]uint64, c.depth)
	for i := uint(0); i < c.depth; i++ {
		hashes[i] = sum1 + uint64(i)*sum2
	}
	return hashes
}

var (
	ErrDimensionMismatch = &errorString{"cannot merge sketches with different dimensions"}
	ErrTooShort         = &errorString{"invalid bytes: too short"}
	ErrInvalidLength    = &errorString{"invalid bytes: incorrect length"}
)

type errorString struct {
	s string
}

func (e *errorString) Error() string {
	return e.s
}